src/Raven.Server/Documents/Indexes/Workers/HandleReferencesBase.ReferencesState.cs

```diff
@@ -0,0 +1,110 @@
+﻿using System.Collections.Generic;
+using Raven.Server.ServerWide.Context;
+
+namespace Raven.Server.Documents.Indexes.Workers
+{
+    public abstract partial class HandleReferencesBase
+    {
+        private class ReferencesState
+        {
+            private readonly Dictionary<string, ReferenceState> _lastIdPerCollectionForDocuments = new Dictionary<string, ReferenceState>();
+
+            private readonly Dictionary<string, ReferenceState> _lastIdPerCollectionForTombstones = new Dictionary<string, ReferenceState>();
+
+            public Reference For(ActionType actionType, string collection)
+            {
+                return new Reference(GetDictionary(actionType), collection);
+            }
+
+            public InMemoryReferencesInfo GetReferencesInfo(string collection)
+            {
+                return new InMemoryReferencesInfo
+                {
+                    ParentItemEtag = new Reference(GetDictionary(ActionType.Document), collection).GetLastIndexedParentEtag(),
+                    ParentTombstoneEtag = new Reference(GetDictionary(ActionType.Tombstone), collection).GetLastIndexedParentEtag(),
+                };
+            }
+
+            public void Clear(ActionType actionType)
+            {
+                var dictionary = GetDictionary(actionType);
+                dictionary.Clear();
+            }
+
+            private Dictionary<string, ReferenceState> GetDictionary(ActionType actionType)
+            {
+                return actionType == ActionType.Document
+                    ? _lastIdPerCollectionForDocuments
+                    : _lastIdPerCollectionForTombstones;
+            }
+
+            public class ReferenceState
+            {
+                public string ReferencedItemId;
+                public string NextItemId;
+                public long ReferencedItemEtag;
+                public long LastIndexedParentEtag;
+            }
+
+            public class Reference
```

Comment:
Resharper warning - Class `Reference` hides class from outer class.

Because in `HandleReferences.cs`, that is partial, you have another:

```
public class Reference
        {
            public LazyStringValue Key;

            public long Etag;
        }
```

Reply:
done